### PR TITLE
use XML instead of annotations for Doctrine mapping to allow overriding it

### LINKED ISSUE #19: Move Entity configuration to separate yml file
As already discussed in #18  the `Setting` entity mapping is coded using annotations, which makes it impossible to override it. So what I'd like to propose is to move from Annotations to a `yml` file so the entity mapping can be overridden in `/app/Resources/CraueConfigBundle`. That way the bundle doesn't change, but allows others to fix this.

### PR DESCRIPTION
**Pull request by craue**
Resolves #19. I've chosen XML over YAML to avoid a new dependency to `symfony/yaml`.

@Rvanlaak: Could you try this to make sure the mapping can actually be overridden as expected?

### CODE REVIEW

**Comment by Rvanlaak** [comment]
> I've chosen XML over YAML to avoid a new dependency to symfony/yaml.

The bundle is already [relying on `symfony/symfony`](https://github.com/craue/CraueConfigBundle/blob/master/composer.json#L27) so that dependency already is available

**Comment by craue** [comment]
> The bundle is already [relying on `symfony/symfony`](https://github.com/craue/CraueConfigBundle/blob/master/composer.json#L27) so that dependency already is available

Only as a dev requirement.

**Comment by craue** [comment]
@Rvanlaak: Fine to merge?

**Comment by Rvanlaak** [comment]
Will check the PR branch later today

**Comment by Rvanlaak** [comment]
Update 1: everything is working fine after switching to the XML annotations.

Will try to override the XML mapping now.

**Comment by Rvanlaak** [comment]
Doctrine tries to drop the UNIQUE index:

```sql
DROP INDEX UNIQ_B95BA9425E237E06 ON craue_config_setting;
```

**Comment by Rvanlaak** [comment]
Next to that, Doctrine won't allow us to override the entity without creating a super-class:
http://symfony.com/doc/current/cookbook/bundles/override.html#entities-entity-mapping
In other words, moving to XML won't solve this problem.

We should make the config class a parameter, in the way how Sylius allows to override models:
http://docs.sylius.org/en/latest/bundles/general/overriding_models.html

**Comment by craue** [comment]
...or introduce a mapped superclass? I've added a commit for that. I now actually like the idea of being able to provide several mappings (ORM, ODM) and using dedicated files (XML or YAML) for that seems most convenient. A class config parameter could be added, but I guess this PR would still be needed in such case. Anyway, this is all new for me, so don't hesitate to prove me wrong. :smirk:

> Doctrine tries to drop the UNIQUE index

Are you sure it's due to this PR? I cannot reproduce that.

**Comment by Rvanlaak** [comment]
> A class config parameter could be added, 

Without that I don't see a way to let your Bundle use the Entity located in my application. Maybe add some docs about extending the `Settings` entity with some example code?

> but I guess this PR would still be needed in such case.

Definitely! :+1:

**Comment by craue** [comment]
I made the class configurable and added a test, but there's one thing I don't have a solution for yet: The mapping file (`Tests/IntegrationTestBundle/Resources/config/doctrine-mapping/CustomSetting.orm.xml`) for the custom (overridden) class needs to be loaded, if one is used. Any idea how to achieve that?

**Comment by Rvanlaak** [comment]
Will pull the changes and take a look into that within the hour :+1:

**Comment by Rvanlaak** [comment]
Hmm, did find the "things to note" in this chapter:

http://doctrine-orm.readthedocs.org/en/latest/reference/inheritance-mapping.html#attribute-override

**Comment by craue** [comment]
Are you pointing to something special? Looks like all this stuff (overriding a model, extending a model, supporting multiple drivers) is not as easy as I thought.

**Comment by Rvanlaak** [comment]
Yeah, the following:

> The column type CANNOT be changed. If the column type is not equal you get a MappingException

**Comment by craue** [comment]
And you want to change the type of the `value` field, right? :smirk:
If so, we cannot use/provide a mapped superclass and have to force users into specifying the complete mapping for the model. I don't like that idea and hope there's another way...

**Comment by Rvanlaak** [comment]
@beberlei is there any chance overriding attribute types will become possible in the near future?

**Comment by Rvanlaak** [comment]
This will probably only work if `value` is moved from the `BaseSetting` to the `Setting` entity. An extra interface is needed to ensure `value` is implemented. 

Then I can determine the way how `value` looks in my own application, without overriding anything.

**Comment by craue** [comment]
> This will probably only work if `value` is moved from the `BaseSetting` to the `Setting` entity.

I tried that locally but still couldn't get the mapping file loaded for the test.

> An extra interface is needed to ensure `value` is implemented.

What do you mean? `SettingInterface` already has methods `setValue` and `getValue`.

**Comment by craue** [comment]
If there's no way to make this work, I'm forced to drop that idea completely.

**Comment by craue** [comment]
@rvanlaak, I started another approach and could finally make it work. 😄 Please take a look.

**Comment by rvanlaak** [comment]
Oh wow really, that would be awesome! So then we finally can make the config a `text` instead of a `varchar` field? ;)

I'm now looking into a nice way to make these configurations translatable / internationalizable :-)

**Comment by craue** [comment]
I just made one more change to render the built-in form fields as `textarea` when the overridden entity maps the field `value` to a `text` column.

> I'm now looking into a nice way to make these configurations translatable / internationalizable :-)

Luckily out of the scope of this PR. 😏

**Comment by craue** [comment]
@rvanlaak, are any further changes needed or do you want me to just push the merge button? 😏

File: composer.json [hunk 1]
```diff
@@ -16,17 +16,17 @@
 		}
 	],
 	"require": {
-		"php": "^5.3.2|~7.0",
-		"symfony/framework-bundle": "~2.3|~3.0"
+		"php": "^5.3.9|~7.0",
```
Comment (rvanlaak) [inline composer.json]: Will this PR release 2.0 of this bundle? If so, what about bumping it to `>=5.5.9`?

File: Entity/SettingInterface.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Craue\ConfigBundle\Entity;
+
+/**
+ * @author Christian Raue <christian.raue@gmail.com>
+ * @copyright 2011-2017 Christian Raue
```
Comment (craue) [inline Entity/SettingInterface.php]: I prefer to keep the year for now. It's not a big deal to perform a search and replace once a year. 😆

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		$treeBuilder->root('craue_config')
+			->children()
+				->scalarNode('db_driver')
+					->defaultValue($supportedDrivers[0])
+					->validate()
+						->ifNotInArray($supportedDrivers)
+						->thenInvalid('The driver "%s" is not supported. Please choose one of ' . json_encode($supportedDrivers))
+					->end()
+				->end()
+				// when at least two values are defined (Symfony < 3.1 throws an exception with only one value)
```
Comment (craue) [inline DependencyInjection/Configuration.php]: Done, added a TODO task instead.

File: composer.json (hunk 1 above)
Comment (craue) [inline composer.json]: I guess 2.0 can be released soon, yes. But I'd like to keep Symfony 2.7 compatibility.

**Pull request merged**